---
title: "GenSEM QC"
author: "Ellen Martin"
date: "07/02/2022"
output: html_document
---

##This QC script includes general data reformatting and clean-up as well as munging using GenomicSEM

## Setting Working Directories

Setting WD to local storage (D://GSEM)
Installing libraries:
- GenomicSEM (v 0.0.2) installed from https://github.com/GenomicSEM/GenomicSEM
- data.table

Setting Output path (outPath) to (D://GSEM//processed GSEM data)

```{r libraries}
filepath="D://GSEM" #main WD
setwd(filepath)
outPath=paste0(filepath, "//processed GSEM data") #output path for saved data

# Load and install libraries
load.lib=c('pheatmap', 'ggthemes', 'ggpubr', 'Qtlizer', 'devtools', 'GenomicSEM', 'data.table', 'Matrix',
           'sem', 'Matrix', 'stats', 'semTools', 'ggcorrplot', 'grex', 'openxlsx', 'rJava', 'qqman', 'gmodels', 'ggplot2',
           'ggcorrplot', 'tidyverse', 'reshape2', 'pdftools', 'plyr', 'phenoscanner','ggpubr',
           'biomaRt', 'viridis', 'rdrop2', 'lavaan', 'gprofiler2', 'knitr',
           'gridExtra', 'CMplot', 'configr', 'purrr', 'TwoSampleMR', 'GOexpress', 'ieugwasr', 'xlsx', 'data.table')


install.lib<-load.lib[!load.lib %in% installed.packages()]

# Install missing packages
for(lib in install.lib) install.packages(lib, dependencies=TRUE)
# Load all packages
sapply(load.lib,require,character=TRUE)
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
## Cleaning and Quality Control of Raw Summary Statistics ##
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Raw Sumstats downloaded either as .gz files or .txt files

```{r qc, echo=FALSE}

##HowardMDD.txt##

MDD_RawSumstats <- fread("D:\\GSEM\\text files\\HowardMDD.txt", header = T) #reading in the RawSumstats using fread()
MDD_Sumstats <- MDD_RawSumstats #keeping the read-in file in case of errors
head(MDD_Sumstats) #checking headers

#changing names to correspond with GenSEM 
names(MDD_Sumstats)[names(MDD_Sumstats)=="MarkerName"] <- "SNP" 
names(MDD_Sumstats)[names(MDD_Sumstats)=="Freq"] <- "MAF"
names(MDD_Sumstats)[names(MDD_Sumstats)=="LogOR"] <- "BETA"
names(MDD_Sumstats)[names(MDD_Sumstats)=="StdErrLogOR"] <- "SE"

#reformatting allele cols
MDD_Sumstats$A1 <- toupper(MDD_Sumstats$A1)
MDD_Sumstats$A2 <- toupper(MDD_Sumstats$A2)

#assigning missing N data
MDD_Sumstats$N <- as.numeric(807553)


#writing a new table of QC'ed sumstats, stored in the outPath
write.table(MDD_Sumstats, paste0(outPath,'//HowardMDD.txt'), col.names=T, row.names=F,quote=F)

##StahlBIP.gz##
BIP_RawSumstats <- fread("D://GSEM//gz files//StahlBIP.gz")
BIP_Sumstats <- BIP_RawSumstats
head(BIP_Sumstats)
BIP_Sumstats <- BIP_Sumstats[,c(1,2,3,4,5,7,8,9,10,11)] #selecting only necessary columns
names(BIP_Sumstats)[names(BIP_Sumstats)=="FRQ_U_31358"] <- "MAF"
BIP_Sumstats$N <- as.numeric(51710)

BIP_Sumstats$BETA <- log(as.numeric(BIP_Sumstats$OR))

#calculate the CI of the OR
BIP_Sumstats$CI_upper <- as.numeric(BIP_Sumstats$OR) + 1.96*BIP$SE
BIP_Sumstats$CI_lower <- as.numeric(BIP_Sumstats$OR) - 1.96*BIP$SE

BIP_Sumstats$SE_B <- as.numeric((log(BIP_Sumstats$CI_upper)-log(abs(BIP_Sumstats$CI_lower)))/(1.96*2))

BIP_Sumstats$P <- as.numeric(BIP_Sumstats$P)

BIP_Sumstats$SE <- BIP_Sumstats$SE_B

write.table(BIP_Sumstats, paste0(outPath,'/StahlBIP.txt'), col.names=T, row.names=F,quote=F)

##OtowaANX.gz##
ANX_RawSumstats <- fread("D://GSEM//gz files//OtowaANX.gz")
ANX_Sumstats <- ANX_RawSumstats
head(ANX_Sumstats)

#renaming
names(ANX_Sumstats)[names(ANX_Sumstats)=="SNPID"] <- "SNP"
names(ANX_Sumstats)[names(ANX_Sumstats)=="Allele1"] <- "A1"
names(ANX_Sumstats)[names(ANX_Sumstats)=="Allele2"] <- "A2"
names(ANX_Sumstats)[names(ANX_Sumstats)=="Freq1"] <- "MAF"
names(ANX_Sumstats)[names(ANX_Sumstats)=="Effect"] <- "BETA"
names(ANX_Sumstats)[names(ANX_Sumstats)=="StdErr"] <- "SE"
names(ANX_Sumstats)[names(ANX_Sumstats)=="P.value"] <- "P"
names(ANX_Sumstats)[names(ANX_Sumstats)=="TotalN"] <- "N"

write.table(ANX_Sumstats, paste0(outPath,'/OtowaANX.txt'), col.names=T, row.names=F,quote=F)

##NievergeltPTSD.gz##
PTSD_RawSumstats <- fread("D://GSEM//gz files//NievergeltPTSD.gz")
PTSD_Sumstats <- PTSD_RawSumstats
head(PTSD_Sumstats)
names(PTSD_Sumstats)[names(PTSD_Sumstats)=="FRQ_A_23212"] <- "MAF"

PTSD_Sumstats <- PTSD_Sumstats[,c(1,2,3,4,5,6,8,9,10,11)]

PTSD_Sumstats$BETA <- log(as.numeric(PTSD_Sumstats$OR))
PTSD_Sumstats$N <- as.numeric(9537)
#calculate the CI of the OR
PTSD_Sumstats$CI_upper <- as.numeric(PTSD_Sumstats$OR) + 1.96*PTSD_Sumstats$SE
PTSD_Sumstats$CI_lower <- as.numeric(PTSD_Sumstats$OR) - 1.96*PTSD_Sumstats$SE

PTSD_Sumstats$SE_B <- as.numeric((log(PTSD_Sumstats$CI_upper)-log(abs(PTSD_Sumstats$CI_lower)))/(1.96*2))

PTSD_Sumstats$P <- as.numeric(PTSD_Sumstats$P)

PTSD_Sumstats$SE <- PTSD_Sumstats$SE_B

write.table(PTSD_Sumstats, paste0(outPath,'/NievergeltPTSD.txt'), col.names=T, row.names=F,quote=F)

##DayLONE.txt##
LONE_RawSumstats <- fread("D://GSEM//text files//DayLONE.txt")
LONE_Sumstats <- LONE_RawSumstats
head(LONE_Sumstats)

#renaming
names(LONE_Sumstats)[names(LONE_Sumstats)=="snpid"] <- "SNP"
names(LONE_Sumstats)[names(LONE_Sumstats)=="chr"] <- "CHR"
names(LONE_Sumstats)[names(LONE_Sumstats)=="bpos"] <- "BP"
names(LONE_Sumstats)[names(LONE_Sumstats)=="a1"] <- "A1"
names(LONE_Sumstats)[names(LONE_Sumstats)=="a2"] <- "A2"
names(LONE_Sumstats)[names(LONE_Sumstats)=="mtag_z"] <- "Z"
names(LONE_Sumstats)[names(LONE_Sumstats)=="n"] <- "N"
names(LONE_Sumstats)[names(LONE_Sumstats)=="mtag_beta"] <- "BETA"
names(LONE_Sumstats)[names(LONE_Sumstats)=="mtag_pval"] <- "P"
names(LONE_Sumstats)[names(LONE_Sumstats)=="freq"] <- "MAF"
names(LONE_Sumstats)[names(LONE_Sumstats)=="mtag_se"] <- "SE"

LONE_Sumstats <- LONE_Sumstats[,c(1,2,3,4,5,7,8,9,10,11,12)] #selecting necessary cols

write.table(LONE_Sumstats, paste0(outPath,'/DayLONE.txt'), col.names=T, row.names=F,quote=F)

##JohnsonCUD.gz##
CAN_RawSumstats <- fread("D://GSEM//gz files//JohnsonCUD.gz")
CAN_Sumstats <- CAN_RawSumstats
head(CAN_Sumstats)

CAN_Sumstats <- CAN_Sumstats[,c(1,2,3,4,5,6,7,8)]

#calculating BETA
CAN_Sumstats$BETA <- CAN_Sumstats$Z / sqrt((2*CAN_Sumstats$P)*(1-CAN_Sumstats$P)*(CAN_Sumstats$N + (CAN_Sumstats$Z)^2))

#calculating SE

CAN_Sumstats$SE <- 1 / sqrt((2*CAN_Sumstats$P)*(1-CAN_Sumstats$P)*(CAN_Sumstats$N + (CAN_Sumstats$Z)^2))

write.table(CAN_Sumstats, paste0(outPath,'/JohnsonCAN.txt'), col.names=T, row.names=F,quote=F)

##SchumannALC.gz##

ALC_RawSumstats <- fread("D://GSEM//gz files//SchumannALC.gz")
ALC_Sumstats <- ALC_RawSumstats
head(ALC_Sumstats)

ALC_Sumstats <- ALC_Sumstats[,c(1,2,3,4,5,6,7,8,9)] #col selection

#renaming
names(ALC_Sumstats)[names(ALC_Sumstats)=="MarkerName"] <- "SNP"
names(ALC_Sumstats)[names(ALC_Sumstats)=="PhysPos"] <- "BP"
names(ALC_Sumstats)[names(ALC_Sumstats)=="Allele1"] <- "A1"
names(ALC_Sumstats)[names(ALC_Sumstats)=="Allele2"] <- "A2"
names(ALC_Sumstats)[names(ALC_Sumstats)=="Freq1"] <- "MAF"
names(ALC_Sumstats)[names(ALC_Sumstats)=="Effect"] <- "BETA"
names(ALC_Sumstats)[names(ALC_Sumstats)=="StdErr"] <- "SE"
names(ALC_Sumstats)[names(ALC_Sumstats)=="P.value"] <- "P"
names(ALC_Sumstats)[names(ALC_Sumstats)=="Chr"] <- "CHR"

#reformatting alleles
ALC_Sumstats$A1 <- toupper(ALC_Sumstats$A1)
ALC_Sumstats$A2 <- toupper(ALC_Sumstats$A2)

write.table(ALC_Sumstats, paste0(outPath,'/SchumannALC.txt'), col.names=T, row.names=F,quote=F)

##WoottonSMK.txt##
SMK_RawSumstats <- fread("D://GSEM//text files//WoottonSMK.txt")
SMK_Sumstats <- SMK_RawSumstats
SMK_Sumstats$N <- as.numeric(462690)

head(SMK_Sumstats) #already QC'ed
write.table(SMK_Sumstats, paste0(outPath,'/WoottonSMK.txt'), col.names=T, row.names=F,quote=F)

##GroveASD##
ASD_RawSumstats <- fread("D://GSEM//GroveASD")
ASD_Sumstats <- ASD_RawSumstats
head(ASD_Sumstats)

names(ASD_Sumstats)[names(ASD_Sumstats)=="EUR_FRQ"] <- "MAF"
ASD_Sumstats$N <- as.numeric(46350)

ASD_Sumstats$BETA <- as.numeric(log(ASD_Sumstats$OR))

ASD_Sumstats$CI_upper <- ASD_Sumstats$OR + 1.96*ASD$SE
ASD_Sumstats$CI_lower <- ASD_Sumstats$OR - 1.96*ASD$SE

ASD_Sumstats$SE_B <- as.numeric((log(ASD_Sumstats$CI_upper)-log(abs(ASD_Sumstats$CI_lower)))/(1.96*2))
ASD_Sumstats$SE <- ASD_Sumstats$SE_B
ASD_Sumstats$P <- as.numeric(ASD_Sumstats$P)

write.table(ASD_Sumstats, paste0(outPath,'/GroveASD.txt'), col.names=T, row.names=F,quote=F)

##DemontisADHD.txt##

ADHD <- fread("D://GSEM//gz files//DemontisADHD.gz")

head(ADHD)
summary(ADHD)

ADHD$N <- ADHD$Nca + ADHD$Nco #adding total N

ADHD$BETA <- as.numeric(log(ADHD$OR)) #converting OR to BETA (log odds)

ADHD$CI_upper <- ADHD$OR + 1.96*ADHD$SE
ADHD$CI_lower <- ADHD$OR - 1.96*ADHD$SE

ADHD$SE_B <- as.numeric((log(ADHD$CI_upper)-log(abs(ADHD$CI_lower)))/(1.96*2))
ADHD$P <- as.numeric(ADHD$P) #formating pval
ADHD$SE <- ADHD$SE_B #replacing SE column with BETA SE
ADHD$MAF <- ADHD$FRQ_A_19099 #MAF creating MAF column

ADHD_clean <- ADHD[,c(1,2,3,4,5,8,10,11,20,21,25)] #selecting relevant columns

ADHD_unique = ADHD_clean[!duplicated(ADHD_clean$SNP)] #remove duplicated SNPs

write.table(ADHD_unique, file = "D:\\GSEM\\processed GSEM data\\GSEMcleansumstats\\DemontisADHD_clean", 
            sep="\t", 
            row.names = FALSE, 
            col.names = TRUE, 
            quote=F)

##PardinasSCZ.txt##

SCZ_RawSumstats <- fread("D://GSEM//text files//PardinasSCZ.txt")
SCZ_Sumstats <- SCZ_RawSumstats
SCZ_Sumstats$N <- as.numeric(105318)
SCZ_Sumstats$BETA <- log(as.numeric(SCZ_Sumstats$OR))

#calculate the CI of the OR
SCZ_Sumstats$CI_upper <- as.numeric(SCZ_Sumstats$OR) + 1.96*SCZ$SE
SCZ_Sumstats$CI_lower <- as.numeric(SCZ_Sumstats$OR) - 1.96*SCZ$SE

SCZ_Sumstats$SE_B <- as.numeric((log(SCZ_Sumstats$CI_upper)-log(abs(SCZ_Sumstats$CI_lower)))/(1.96*2))

SCZ_Sumstats$P <- as.numeric(SCZ_Sumstats$P)

SCZ_Sumstats$SE <- SCZ_Sumstats$SE_B
head(SCZ_Sumstats) #aready QC'ed

write.table(SCZ_Sumstats, file="PardinasSCZ.txt", 
            sep="\t", 
            row.names = FALSE, 
            col.names = TRUE, 
            quote=F)
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
## Munging using GenomicSEM ##
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r munge, echo=FALSE}

setwd(outPath) #sets new WD to outPath so cleaned and munged files are saved

##munging
files <- c("DemontisADHD.txt","OtowaANX.txt","GroveASD.txt","StahlBIP.txt","JohnsonCAN.txt","SchumannALC.txt","HowardMDD.txt","PardinasSCZ.txt","NievergeltPTSD.txt","WoottonSMK.txt","DayLONE.txt")
trait.names<-c("ADHD","ANX","ASD","BIP","CAN","ALC","MDD","SCZ","PTSD","SMK","LONELINESS")
hm3 = "D://GSEM//w_hm3.noMHC.snplist" #hapmap3 reference alleles downloaded 

#N = total sample size for traits
N<-c(55374,17310,46350,51710,384032,70460,807553,105318,9537,462690,487647)

#Run the munge function. This will create all the .sumstats.gz files (e.g., DEP_HM3.sumstats.gz)
munge(files=files,hm3=hm3, trait.names=trait.names, N=N)

```

